### DISCUSSION TITLE
Interplay between `reservationConcurrency`  and `discoverRelays`

### DISCUSSION THREAD

**Discussion by 2color** [libp2p/js-libp2p#2654] [Q&A] [answered]
While experimenting with `circuitRelayTransport` configuration, I found that js-libp2p will not to create a reservation with the following configuration, even though the peer is already connected to circuit relay peer: 

```ts
circuitRelayTransport({
    discoverRelays: 0,
    reservationConcurrency: 1
}),
```

Only setting `discoverPeers` to 1, results in a reservation.
```ts
circuitRelayTransport({
    discoverRelays: 1,
    reservationConcurrency: 1
}),
```

What's weird is if I connect to two relays manually (via a bootstrap list), set `discoverRelays: 2` and `reservationConcurrency: 1`, the browser peer will get a reservations with both. So the logic seems to be the "highest of the two values":

```
const libp2p = await createLibp2p({
    metrics: devToolsMetrics(),
    addresses: {
      listen: [
        // 👇 Listen for webRTC connection
        '/webrtc',
      ],
    },
    transports: [
      webSockets({
        // Allow all WebSocket connections inclusing without TLS
        filter: filters.all,
      }),
      webTransport(),
      webRTC(),
      // // 👇 Required to create circuit relay reservations in order to hole punch browser-to-browser WebRTC connections
      circuitRelayTransport({
        discoverRelays: 2,
        reservationConcurrency: 1
      }),
    ],
    connectionEncryption: [noise()],
    streamMuxers: [yamux()],
    connectionGater: {
      // Allow private addresses for local testing
      denyDialMultiaddr: async () => false,
    },
    peerDiscovery: [
      bootstrap({
        list: [
          '/ip4/127.0.0.1/tcp/9001/ws/p2p/12D3KooWA7QLGSgDNB3NdYmHEdEDU3FHv3PzjW9mHsEJst4e2Rqi',
          '/ip4/127.0.0.1/tcp/9003/ws/p2p/12D3KooWDJ9J5PVyxgqKg93wqaA7azr525YWakQGeYS9hRvZyCK6'
        ],
      }),
    ],
    services: {
      pubsub: gossipsub(),
      identify: identify(),
    },
  })
```

---

Is this a bug? or expected behaviour? It seems to me that setting `reservationConcurrency: 1` should limit the reservations.

**Comment by achingbrain** [comment, marked as answer]
These settings control different things.

`discoverRelays` means "Make reservations on this many relays".  So setting it to `0` means you won't make any reservations, setting it to `100` means you'll make 100 reservations.

`reservationConcurrency` means "Try to make this many reservations at the same time".  So setting it to `1` will make reservations on relays sequentially (bad if you encounter a relay that's slow to respond), setting it to a greater number will make that many reservation attempts in parallel, but the total number of reservations you'll make is still controlled by `discoverRelays`.
